LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -64,514 +65,541 @@ public class DefaultLoRaDataRequestHandler : ILoRaDataRequestHandler
 
             List<Task> deferredTasks = null;
+            IAsyncDisposable deviceConnectionActivity = null;
 
-            var timeWatcher = request.GetTimeWatcher();
-
-            var loraPayload = (LoRaPayloadData)request.Payload;
-            this.d2cPayloadSizeHistogram?.Record(loraPayload.Frmpayload.Length);
-
-            var payloadFcnt = loraPayload.Fcnt;
+            try
+            {
+                ExceptionDispatchInfo edi = null;
+                LoRaDeviceRequestProcessResult result = null;
 
-            var payloadFcntAdjusted = LoRaPayloadData.InferUpper32BitsForClientFcnt(payloadFcnt, loRaDevice.FCntUp);
-            this.logger.LogDebug($"converted 16bit FCnt {payloadFcnt} to 32bit FCnt {payloadFcntAdjusted}");
+                try
+                {
+                    result = await ProcessAsync();
+                }
+#pragma warning disable CA1031 // Do not catch general exception types (exception is rethrown)
+                catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                {
+                    edi = ExceptionDispatchInfo.Capture(ex);
+                }
 
-            var requiresConfirmation = loraPayload.RequiresConfirmation;
+                Exception deferredTaskException = null;
+                if (deferredTasks is { } someDeferredTasks)
+                {
+                    _ = await Task.WhenAny(Task.WhenAll(someDeferredTasks));
+                    var exs = someDeferredTasks.GetExceptions();
 
-            LoRaADRResult loRaADRResult = null;
+                    deferredTaskException = exs switch
+                    {
+                        { Count: 1 } => exs[0],
+                        { Count: > 1 } => new AggregateException(exs),
+                        _ => null
+                    };
+                }
 
-            var frameCounterStrategy = this.frameCounterUpdateStrategyProvider.GetStrategy(loRaDevice.GatewayID);
-            if (frameCounterStrategy == null)
-            {
-                this.logger.LogError($"failed to resolve frame count update strategy, device gateway: {loRaDevice.GatewayID}, message ignored");
-                return new LoRaDeviceRequestProcessResult(loRaDevice, request, LoRaDeviceRequestFailedReason.ApplicationError);
-            }
+                if (edi is { } someEdi)
+                {
+                    if (deferredTaskException is { } someDeferredTaskException)
+                        this.logger.LogError(someDeferredTaskException, "Processing of deferred tasks failed.");
 
-            // Contains the Cloud to message we need to send
-            IReceivedLoRaCloudToDeviceMessage cloudToDeviceMessage = null;
+                    someEdi.Throw();
+                }
+                else if (deferredTaskException is { } someDeferredTaskException)
+                {
+                    throw someDeferredTaskException;
```

Comment:
Do we want to fail the data request processing if any of the forked tasks failed? This means that if data request processing succeeded then we'll mask it with a errors from forked tasks.

Reply:
Changed with 8a82a1f3a5182e059c5e29f4cc08c7495d6eb485, will also update the PR description